File: python/providers/claude.py
import os
from typing import Dict, Any, List, Optional
from .base import Provider
import sys
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))
from response_parser import ResponseParser


class ClaudeProvider(Provider):
    """Claude API provider using Anthropic API"""

    def __init__(self):
        self.api_key = os.getenv('ANTHROPIC_API_KEY')
        self.anthropic = None
        self.name = 'anthropic'  # For parser hints
        self.parser = ResponseParser()

        # Initialize Anthropic client
        if self.api_key:
            try:
                import anthropic
                self.anthropic = anthropic.Anthropic(api_key=self.api_key)
            except ImportError:
                raise Exception(
                    "anthropic package not installed. Run: pip install anthropic")
        else:
            raise Exception(
                "ANTHROPIC_API_KEY environment variable not set")


    async def complete(
        self,
        instruction: str,
        context: str,
        model: str = 'claude-3-5-sonnet-20241022',
        temperature: float = 0.7,
        max_tokens: int = 4096,
        **kwargs
    ) -> Dict[str, Any]:
        """Generate completion using Claude"""
        prompt = self.build_prompt(instruction, context)

        if not self.anthropic:
            raise Exception("Claude API key not configured")

        try:
            response = self.anthropic.messages.create(
                model=model,
                max_tokens=max_tokens,
                temperature=temperature,
                messages=[
                    {"role": "user", "content": prompt}
                ]
            )

            content = response.content[0].text if response.content else ""
            return self.parse_response(content, provider_hint='anthropic')

        except Exception as e:
            raise Exception(f"Claude API error: {str(e)}")

    async def chat(
        self,
        messages: List[Dict[str, str]],
        model: str = 'claude-3-5-sonnet-20241022',
        temperature: float = 0.7,
        max_tokens: int = 4096,
        **kwargs
    ) -> Dict[str, Any]:
        """Handle chat with Claude"""
        if not self.anthropic:
            raise Exception("Claude API key not configured")

        # Convert messages to Claude format
        claude_messages = []
        for msg in messages:
            # Skip system messages or convert them to user messages with context
            if msg['role'] == 'system':
                claude_messages.append({
                    "role": "user",
                    "content": f"Context: {msg['content']}"
                })
            else:
                claude_messages.append({
                    "role": msg['role'] if msg['role'] != 'ai' else 'assistant',
                    "content": msg['content']
                })

        try:
            response = self.anthropic.messages.create(
                model=model,
                max_tokens=max_tokens,
                temperature=temperature,
                messages=claude_messages
            )

            content = response.content[0].text if response.content else ""

            # Check if response contains code
            parsed = self.parse_response(content, provider_hint='anthropic')
            if parsed.get('code'):
                return parsed
            else:
                return {"content": content}

        except Exception as e:
            raise Exception(f"Claude API error: {str(e)}")


    def get_available_models(self) -> List[str]:
        """Return list of available Claude models"""
        return [
            'claude-3-5-sonnet-20241022',  # Latest Sonnet 3.5
            'claude-3-5-haiku-20241022',   # Latest Haiku 3.5
            'claude-3-opus-20240229',      # Opus 3.0
            'claude-3-sonnet-20240229',    # Sonnet 3.0
            'claude-3-haiku-20240307',     # Haiku 3.0
        ]

File: rust/src/providers/claude.rs
use async_trait::async_trait;
use serde_json::json;
use std::collections::HashMap;

use crate::config::Config;
use crate::http;
use crate::retry::{self, RetryConfig};

use super::{ChatMessage, Provider, ProviderResult};

pub struct ClaudeProvider;

const API_URL: &str = "https://api.anthropic.com/v1/messages";

fn get_api_key() -> Result<String, String> {
    std::env::var("ANTHROPIC_API_KEY")
        .map_err(|_| "ANTHROPIC_API_KEY not set".to_string())
}

fn build_headers(api_key: &str) -> HashMap<String, String> {
    let mut headers = HashMap::new();
    headers.insert("content-type".to_string(), "application/json".to_string());
    headers.insert("x-api-key".to_string(), api_key.to_string());
    headers.insert("anthropic-version".to_string(), "2023-06-01".to_string());
    headers
}

fn extract_content(body: &serde_json::Value) -> Result<String, String> {
    // Check for API errors
    if let Some(error) = body.get("error") {
        let msg = error
            .get("message")
            .and_then(|m| m.as_str())
            .unwrap_or("unknown error");
        return Err(format!("Claude API error: {msg}"));
    }

    // Extract content from response
    body.get("content")
        .and_then(|c| c.as_array())
        .and_then(|arr| arr.first())
        .and_then(|item| item.get("text"))
        .and_then(|t| t.as_str())
        .map(|s| s.to_string())
        .ok_or_else(|| format!("No content in Claude response: {body}"))
}

#[async_trait]
impl Provider for ClaudeProvider {
    async fn complete(
        &self,
        system_prompt: &str,
        user_prompt: &str,
        config: &Config,
    ) -> Result<ProviderResult, String> {
        let api_key = get_api_key()?;
        let headers = build_headers(&api_key);

        let body = json!({
            "model": config.model,
            "max_tokens": config.max_tokens,
            "temperature": config.temperature,
            "system": system_prompt,
            "messages": [{"role": "user", "content": user_prompt}]
        });

        let timeout = config.timeouts.llm_request;
        let retry_config = RetryConfig::default();

        let response = retry::with_retry(&retry_config, || {
            let headers = headers.clone();
            let body = body.clone();
            async move {
                http::request(API_URL, "POST", &headers, Some(&body), timeout).await
            }
        })
        .await?;

        let content = extract_content(&response.body)?;
        Ok(ProviderResult {
            content,
            provider_hint: "claude".to_string(),
        })
    }

    async fn chat(
        &self,
        system_prompt: &str,
        messages: &[ChatMessage],
        config: &Config,
    ) -> Result<ProviderResult, String> {
        let api_key = get_api_key()?;
        let headers = build_headers(&api_key);

        // Convert messages to Claude format
        let claude_messages: Vec<serde_json::Value> = messages
            .iter()
            .map(|msg| {
                let role = match msg.role.as_str() {
                    "system" => "user",
                    "ai" => "assistant",
                    other => other,
                };
                let content = if msg.role == "system" {
                    format!("Context: {}", msg.content)
                } else {
                    msg.content.clone()
                };
                json!({"role": role, "content": content})
            })
            .collect();

        let body = json!({
            "model": config.model,
            "max_tokens": config.max_tokens,
            "temperature": config.temperature,
            "system": system_prompt,
            "messages": claude_messages
        });

        let timeout = config.timeouts.llm_request;
        let retry_config = RetryConfig::default();

        let response = retry::with_retry(&retry_config, || {
            let headers = headers.clone();
            let body = body.clone();
            async move {
                http::request(API_URL, "POST", &headers, Some(&body), timeout).await
            }
        })
        .await?;

        let content = extract_content(&response.body)?;
        Ok(ProviderResult {
            content,
            provider_hint: "claude".to_string(),
        })
    }
}
